### PR TITLE
[#6497] Require icp -f -R to target resource with replica (main)

### PR DESCRIPTION
**Pull request by alanking**
Addresses #6497 

New tests all work for federation, core, and topology (provider and consumer). Running the rest of the tests now

### CODE REVIEW

File: scripts/core_tests_list.json [hunk 1]
```diff
@@ -33,7 +33,8 @@
     "test_icommands_file_operations.Test_ICommands_File_Operations_3",
     "test_icommands_file_operations.Test_ICommands_File_Operations_4",
     "test_icommands_file_operations.Test_ICommands_File_Operations_5",
-    "test_icp",
+    "test_icp.Test_Icp",
+    "test_icp.test_icp_overwrite_with_target_resource",
```
Comment (trel) [inline scripts/core_tests_list.json:37]: is this correct?   these two lines look different... capitalization, etc...

File: scripts/irods/test/test_icp.py [hunk 2]
```diff
@@ -218,3 +218,181 @@ def test_icp_with_apostrophe_logical_path__issue_5759(self):
...
+            self.user.assert_icommand(['irm', '-f', other_logical_path])
+
+    def test_overwrite_fails_when_target_resource_has_no_replica__issue_6497(self):
+        object_name = 'test_overwrite_fails_when_target_resource_has_no_replica__issue_6497'
+        other_object_name = 'test_overwrite_fails_when_target_resource_has_no_replica__issue_6497_other'
+        logical_path = os.path.join(self.user.session_collection, object_name)
+        other_logical_path = os.path.join(self.user.session_collection, other_object_name)
+
+        original_content = 'goin down to cowtown'
+        new_content = 'gonna see the cow beneath the sea'
```
Comment (trel) [inline scripts/irods/test/test_icp.py:295]: duplicate poetry alert - need different words here.

**Review by trel — COMMENTED** [review]
very impressive.

File: scripts/core_tests_list.json (hunk 1 above)
Comment (alanking) [reply scripts/core_tests_list.json:37]: Whups, I did rename the class. But the first one is correct. Second one should be `test_icp.test_overwrites`.

File: scripts/irods/test/test_federation.py [hunk 3]
```diff
@@ -1943,3 +1943,222 @@ def test_iget_data_object_as_user_with_null_access_and_replica_only_in_archive__
...
+
+class test_icp_overwrite_with_target_resource(unittest.TestCase):
+	@classmethod
+	def setUpClass(self):
+		# Create a session as the administrator for the remote zone so we can do things in the remote zone.
+		self.remote_admin = session.make_session_for_existing_user(
+			test.settings.PREEXISTING_ADMIN_PASSWORD,
+			test.settings.PREEXISTING_ADMIN_PASSWORD,
+			test.settings.FEDERATION.REMOTE_HOST,
+			test.settings.FEDERATION.REMOTE_ZONE)
```
Comment (korydraughn) [inline scripts/irods/test/test_federation.py:1956]: Is this correct?

Shouldn't the first argument be the username?

**Review by korydraughn — COMMENTED** [review]
Very nice.

File: scripts/irods/test/test_federation.py (hunk 3 above)
Comment (alanking) [reply scripts/irods/test/test_federation.py:1956]: Well, you'd think that, but there is no such thing as `PREEXISTING_ADMIN_PASSWORD` (currently). I can add that and update this where this situation occurs. I think I lifted this from another test I wrote before...

File: scripts/irods/test/test_icp.py (hunk 2 above)
Comment (alanking) [reply scripts/irods/test/test_icp.py:295]: Right-o

File: scripts/core_tests_list.json (hunk 1 above)
Comment (alanking) [reply scripts/core_tests_list.json]: Done

File: scripts/irods/test/test_icp.py (hunk 2 above)
Comment (alanking) [reply scripts/irods/test/test_icp.py]: Done. Ran out of TMBG lyrics that live in my head :)

File: scripts/irods/test/test_federation.py (hunk 3 above)
Comment (alanking) [reply scripts/irods/test/test_federation.py]: Done

Comment (korydraughn) [reply scripts/irods/test/test_federation.py]: Sure, but does it actually pass with that? Does it not care about the username?

Comment (alanking) [reply scripts/irods/test/test_federation.py]: The username for the admin is `rods` and the password is `rods`, so this works as-is. I went ahead and added the username as a separate item should we change this in the future.

Comment (korydraughn) [reply scripts/irods/test/test_federation.py]: Ohhhh. Well that explains it.

Having an actual variable for `PREEXISTING_ADMIN_USERNAME` makes sense, even if it's identical to the password.

That can happen later too. Up to you.

File: scripts/irods/test/test_icp.py [hunk 4]
```diff
@@ -291,8 +291,8 @@ def test_overwrite_fails_when_target_resource_has_no_replica__issue_6497(self):
         logical_path = os.path.join(self.user.session_collection, object_name)
         other_logical_path = os.path.join(self.user.session_collection, other_object_name)
 
-        original_content = 'goin down to cowtown'
-        new_content = 'gonna see the cow beneath the sea'
+        original_content = 'you will see this'
+        new_content = 'and you will not see this'
```
Comment (trel) [inline scripts/irods/test/test_icp.py:295]: i propose...

```
You're older than you've ever been
And now you're even older,
```

Comment (alanking) [reply scripts/irods/test/test_icp.py]: It is done

**Comment by trel** [comment]
looking good.  squash when ready.

**Comment by alanking** [comment]
Getting consistent failure of `test_delay_queue.Test_Delay_Queue.test_exception_in_delay_server`... Trying again with tip of main to see if it's my change (that would be highly surprising).

**Comment by alanking** [comment]
Seems that it's not just my change. Created an issue for this here: https://github.com/irods/irods/issues/7412

**Comment by alanking** [comment]
Squashed. All tests besides those noted above are passing.

**Comment by korydraughn** [comment]
Oh, just saw your message about delay queue test failures.

Waiting to hear what you find in regard to that.

**Comment by alanking** [comment]
The delay queue failures seem unrelated to my changes as the failure occurs consistently with a build on the tip of main. Since I've already created an issue for that, I think this is good to go. Thoughts?

**Comment by trel** [comment]
makes sense to me.

**Comment by korydraughn** [comment]
Ship it.

**Pull request merged**